kytos/core/interface.py

```diff
@@ -173,19 +177,295 @@ def status_reason(self):
+        Simulates sliding window O(n+m)"""
+        available = self.available_tags[tag_type]
+        if not available:
+            return "Every tag is used."
+
+        error_msg = self.verify_first_last_tag(tag_ranges)
+        if error_msg:
+            return error_msg
+
+        # Corner case
+        ava_n = len(available)
+        if ava_n == 1:
+            return None
+
+        rest_n = len(tag_ranges)
+        ava_i, rest_i = 0, 0
+        while rest_i < rest_n:
+            if tag_ranges[rest_i][0] <= available[ava_i][1] + 1:
+                while ava_i + 1 < ava_n:
+                    if tag_ranges[rest_i][1] + 1 >= available[ava_i+1][0]:
+                        ava_i += 1
+                    elif tag_ranges[rest_i][1] <= available[ava_i][1]:
+                        break
+                    else:
+                        reason = f"Gap detected available_tag:"\
+                                 f"{[available[ava_i], available[ava_i+1]]}"\
+                                 f" and tag_ranges:{tag_ranges[rest_i]}."
+                        return reason
+                rest_i += 1
+            else:
+                reason = f"Gap detected available_tag:"\
+                         f"{[available[ava_i], available[ava_i+1]]} and "\
+                         f"tag_ranges:{tag_ranges[rest_i]}."
+                return reason
+        return None
+
+    def set_tag_ranges(self, tag_ranges: list[list[int]], tag_type: str):
+        """Set new restriction"""
+        with self._tag_lock:
+            reason = self.not_applicable(tag_ranges, tag_type)
+            if reason:
+                msg = f"tag_ranges: {tag_ranges} is not applicable. {reason}"
+                raise KytosResizingAvailableTagError(msg)
+            self.tag_ranges[tag_type] = tag_ranges
+
+    def remove_tag_ranges(self, tag_type: str = '1'):
+        """Remove restriction set on tag_ranges"""
+        self.tag_ranges[tag_type] = None
+
+    def get_next_available_tag(self) -> Optional[int]:
+        """Get the next available tag from the interface.
+
+        Return the next available tag if exists and remove from the
+        available tags.
+        If no tag is available return False.
+        """
+        try:
+            with self._tag_lock:
+                first_tag = self.available_tags[-1][1]
+                self.available_tags[-1][1] -= 1
+                if self.available_tags[-1][0] > self.available_tags[-1][1]:
+                    self.available_tags.pop()
+                return first_tag
+        except IndexError:
+            return False
+
+    @staticmethod
+    def find_index_remove(
+        available_tags: list[list[int]],
+        tag_range: list[int]
+    ) -> Optional[int]:
+        """Find the index of tags in available_tags to be removed"""
+        low = 0
+        high = len(available_tags)
+        while low < high:
+            mid = (low+high)//2
+            if available_tags[mid][0] <= tag_range[0]:
```

Comment:
It'd expect `bisect.bisect_right` to work here too if you could explore it. Although  when removing a `tag_range`, from what I followed, you probably want `bisect_left` since you're actually trying to search the matched index, and if index < len(available_tags) and available_tags[index] overlaps with tag_range then you found it.

Reply:
Now using `bisect.bisect_left` here as well